syntax = "proto2";

package proto;

message ReflectionalSymmetryDetectionOptions {
  // Voxel size used in voxelgrid downsample.
  optional float voxel_size = 1 [default = 0.0];
  
  // Controls number of initial reflectional symmetry candidate proposals.
  // Higher value - higher chance of reconvering correct symmetry at the cost of
  // more false positives and longer runtime.
  optional int32 num_angle_divisions = 2 [default = 5];
    
  optional float flatness_threshold = 3 [default = 0.005];
  
  // Max number of optimization iterations used to refine the candidate
  // symmetries.
  optional int32 refine_iterations = 4 [default = 20];
  
  // Maximum allowed distance between the reflections of two points forming a
  // symmetric correspondence.
  optional float max_correspondence_reflected_distance = 5 [default = 0.01];
  
  // Occlusion distances below this value are clamped to 0.
  optional float min_occlusion_distance = 6 [default = 0.01];
  
  // Occlusion distances above this value are clamped to 1.
  optional float max_occlusion_distance = 7 [default = 0.2];
  
  // A symmetric correspondence with a correspondence angle lower than this has
  // the highest symmetric correspondence score i.e. 0.  
  optional float min_inlier_normal_angle_deg = 8 [default = 10.0];
  
  // A symmetric correspondence with a correspondence angle greater than this
  // has the lowest symmetric correspondence score i.e. 1.  
  optional float max_inlier_normal_angle_deg = 9 [default = 15.0];

  // Maximum allowed occlusion score for a valid symmetry.
  optional float max_occlusion_score = 10 [default = 0.01];
  
  // Minimum allowed symmetric correspondence score normalized by the total
  // number of points in the cloud. Scale [0, 1].
  optional float min_cloud_inlier_score = 11 [default = 0.2];
  
  // Minimum allowed symmetric correspondence score normalized by the total
  // number of symmetric correspondences. Scale [0, 1].  
  optional float min_corresp_inlier_score = 12 [default = 4.0];
  
  // Two symmetries are considered similar if the angle between their normals is
  // less than this threshold.
  optional float symmetry_min_angle_diff_deg = 13 [default = 7.0];  
  
  // Two symmetries are considered similar if the difference between their
  // distances to origin is less than this threshold.
  optional float symmetry_min_distance_diff = 14 [default = 0.02];
  
  // Maximum distance between the reference points of two symmetries that can be
  // merged.
  optional float max_reference_point_distance = 15 [default = 0.3];
}